pilot/pkg/model/service.go

```diff
@@ -488,11 +482,30 @@ func (h Hostname) String() string {
 //  Hostname("*.com").Matches("foo.com")     = true
 //  Hostname("*.foo.com").Matches("foo.com") = false
 func (h Hostname) Matches(o Hostname) bool {
-	if !strings.Contains(string(h), "*") && !strings.Contains(string(o), "*") {
+	if len(h) == 0 && len(o) == 0 {
```

Comment:
Clean this up a bit to something like (removing the newlines, removing the duplicate comparisons to asterisks):

```go
func (h Hostname) Matches(o Hostname) bool {
	if len(h) == 0 && len(o) == 0 {
		return true
	}

	hWildcard := string(h[0]) == '*'
	if hWildcard && len(o) == 0 {
		return true
	}

	oWildcard := string(o[0]) == '*'
	if !hWildcard && !oWildcard {
                 // both are non-wildcards, so do normal string comparison
		return h == o
	}

	longer, shorter := string(h), string(o)
	if hWildcard {
		longer = string(h[1:])
	}
	if oWildcard {
		shorter = string(o[1:])
	}
	if len(longer) < len(shorter) {
		longer, shorter = shorter, longer
		hWildcard, oWildcard = oWildcard, hWildcard
	}

	matches := strings.HasSuffix(longer, shorter)
	if matches && hWildcard && !oWildcard && strings.TrimSuffix(longer, shorter) == "." {
		// we match, but the longer is a wildcard and the shorter is not; we need to ensure we don't match input
		// like `*.foo.com` to `foo.com` in that case (to avoid matching a domain literal to a wildcard subdomain)
		return false
	}
	return matches
}
```

Reply:
let me benchmark this and see what we get (should still be fine).

Reply:
this went from a bench of 21.6 ns/op to 20.5 ns/op with that change (so a tiny savings but still good). That was with a one result test, going to change the bench to make it work a little harder.